### DISCUSSION TITLE
Correct way to seed relations

### DISCUSSION THREAD

**Discussion by matt-chaib** [prisma/prisma#20896] [Q&A] [answered]
### Question

Hi Prisma folk,

In my schema I have "Events", which can take on a "Collision" relation. Likewise, a "Collision" is related to a "CollisionSeverity" model.

```Prisma
model Event {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  longitude Float
  latitude  Float

  markerTypeKey MarkerType @relation(fields: [markerEventType], references: [markerEventTypeDetail]) // 1: Collision 2: Near-Miss 3: Aggression

  markerEventType String

  description String

  timeOfEvent DateTime @default(now())

  Collision  Collision?
  NearMiss   NearMiss?
  Aggression Aggression?
}
```

```Prisma
model Collision {
  id               Int               @id @default(autoincrement())
  severity         CollisionSeverity @relation(fields: [severityType], references: [severityType])
  severityType     String
  estimatedCost    Int
  event            Event             @relation(fields: [eventId], references: [id])
  eventId          Int               @unique
}
```

```Prisma
model CollisionSeverity {
  id           Int         @id @default(autoincrement())
  severityType String      @unique // Fatal, Serious, Moderate, Minor
  collisions   Collision[]
}
```

Each Event has a one-to-one relation to a Collision, NearMiss or Aggression (in this example I'm just trying to get it working for the collision relation). Likewise, a Collision entity should only relate to one Event. A CollisionSeverity entity can map to many Collision entities.

If the above schema is correct, what is the best way to seed for example, three events, each of which is a collision, of severity "Serious"?

I've been trying like this:

```Typescript

let defaultMarkers = [
    { lat: 53.38, lng: -1.47, markerEventType: "Near-Miss"},
    { lat: 53.374, lng: -1.479, markerEventType: "Collision"},
    { lat: 53.38, lng: -1.484, markerEventType: "Aggression" }
]

let collisionSeverityTypes = [
    "Fatal",
    "Serious",
    "Moderate",
    "Mild"
]

let collisions = [
    {
        severity: {connect: { id: 2 }},
        estimatedCost: 100,
        event: {connect: { id: 1 }}
    }
]

const seedMarkers = () => {
    collisionSeverityTypes.forEach(async entry => {
        console.log(entry)
        await db.collisionSeverity.create({
            data: {
                severityType: entry
            }
        })
    })

    collisions.forEach(async collision => {
        await db.collision.create({data: collision })
    }) 

    defaultMarkers.forEach(async m => {
        if (m.markerEventType == "Collision") {
            await db.event.create({data: {
                latitude: m.lat,
                longitude: m.lng,
                markerEventType: m.markerEventType,
                description: "",
                Collision: { connect: { id: 1 } }
            }})
        }}
}

seedMarkers()
```

But I get errors telling me that CollisionSeverity has not been created yet (so when I try to create a collision object it fails), and then even if I create these manually, I get an error telling me "No Event Record found for nested connect...". 

It feels like a chicken-and-egg, because the "Collision" record links to an "Event" record, but the "Event" also links to a "Collision", if you see what I mean? Which do I create first? I was wondering therefore if it's an issue with how I've laid out my schema, or if I am not seeding my database in the correct way.

Thanks,
Matt

### Expected behavior (optional)

Database should seed with three markers, one of which is linked to a collision model which the correct collision severity.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @matt-chaib 👋 

Thank you for raising this question.

From your code
```js

let collisions = [
    {
        severity: {connect: { id: 2 }},
        estimatedCost: 100,
        event: {connect: { id: 1 }}
    }
]
 collisions.forEach(async collision => {
        await db.collision.create({data: collision })
    })
```

You are creating a collision and trying to connect to an event but no event has been created yet. You will need to have at least one `Event` created for linking with `Collision`. 

Also you defined the `collisions` array to have `severity: {connect: { id: 2 }}` but the `severity` field references `severityType` and not `id` as defined in your Schema.

```prisma
severity         CollisionSeverity @relation(fields: [severityType], references: [severityType])
```

You can update your `seedMarkers` function to look like:
```js
const seedMarkers = async () => {
  // Step 1: Seed CollisionSeverity
  for (const severityType of collisionSeverityTypes) {
      await prisma.collisionSeverity.create({
          data: {
              severityType: severityType
          }
      })
  }

  // Step 2: Seed Events
  for (const marker of defaultMarkers) {
      if (marker.markerEventType === "Collision") {
          await prisma.event.create({
              data: {
                  latitude: marker.lat,
                  longitude: marker.lng,
                  markerEventType: marker.markerEventType,
                  description: ""
              }
          })
      }
  }

  // Get all created events
  const events = await prisma.event.findMany();

  // Step 3: Seed Collisions and link them to Events
  for (const collision of collisions) {
      await prisma.collision.create({
          data: {
              severity: { connect:   { severityType: collision.severity.severityType } } ,
              estimatedCost: collision.estimatedCost,
              event: { connect: { id: events[0].id } } // Connect to the first event, for example
          }
      });
  }
}
```

I also update collisions array to look like:
```js
let collisions = [
  {
      severity: { severityType: "Fatal" },
      estimatedCost: 100,
      event: {connect: { id: 1 }}
  }
]
```
With these changes, your seed should work as expected.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludralph** [reply]
Hi @matt-chaib 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by nikolasburk** [reply]
@matt-chaib Did you intentionally not use an [`enum`](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#enum) for modeling `CollisionSeverity` in your schema? It seems like this could notably simplify the schema and the seeding would also become easier:

```prisma
model Collision {
  id               Int               @id @default(autoincrement())
  severity         CollisionSeverity
  severityType     String
  estimatedCost    Int
  event            Event             @relation(fields: [eventId], references: [id])
  eventId          Int               @unique
}

enum CollisionSeverity {
  Fatal
  Serious
  Moderate
  Minor
}
```
